### PR TITLE
Fix error message in variable definition check

### LINKED ISSUE #2312: var has not key 'diagnostic' error when checking var from the recipe
**Describe the bug**
When a necessary entry is missing in a variable, instead of raising the correct error defined in
https://github.com/ESMValGroup/ESMValCore/blob/989839f596a6b4433731c0efcb6b5cb4f3ca7045/esmvalcore/_recipe/check.py#L103-L112

another error is raised as `var` has no key `'diagnostic`. Also a format `f` is missing in line 111.

Solving it with a PR.

**Please attach**
An example with missing `mip` won't produce the expected error message.
```yml
datasets:
- project: OBS
  institute: ukmo
  type: obs
  dataset: en4-v4.2.2
  freq_base: ''
  freq_folder: monthly_mean
  timerange: 190001/190001
diagnostics:
  temperature:
    variables:
      thetao: {}
    scripts: null
```

```
2024-01-30 16:30:49,330 UTC [3133476] ERROR   Program terminated abnormally, see stack trace below for more information:
Traceback (most recent call last):
  File "/gpfs/projects/bsc32/software/suselinux/11/software/ESMValTool/2.10.0/lib/python3.11/site-packages/esmvalcore/_main.py", line 518, in run
    fire.Fire(ESMValTool())
  File "/gpfs/projects/bsc32/software/suselinux/11/software/ESMValTool/2.10.0/lib/python3.11/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/projects/bsc32/software/suselinux/11/software/ESMValTool/2.10.0/lib/python3.11/site-packages/fire/core.py", line 475, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
                                ^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/projects/bsc32/software/suselinux/11/software/ESMValTool/2.10.0/lib/python3.11/site-packages/fire/core.py", line 691, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/projects/bsc32/software/suselinux/11/software/ESMValTool/2.10.0/lib/python3.11/site-packages/esmvalcore/_main.py", line 405, in run
    self._run(recipe, session)
  File "/gpfs/projects/bsc32/software/suselinux/11/software/ESMValTool/2.10.0/lib/python3.11/site-packages/esmvalcore/_main.py", line 447, in _run
    process_recipe(recipe_file=recipe, session=session)
  File "/gpfs/projects/bsc32/software/suselinux/11/software/ESMValTool/2.10.0/lib/python3.11/site-packages/esmvalcore/_main.py", line 127, in process_recipe
    recipe = read_recipe_file(recipe_file, session)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/projects/bsc32/software/suselinux/11/software/ESMValTool/2.10.0/lib/python3.11/site-packages/esmvalcore/_recipe/recipe.py", line 73, in read_recipe_file
    return Recipe(raw_recipe, session, recipe_file=filename)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/projects/bsc32/software/suselinux/11/software/ESMValTool/2.10.0/lib/python3.11/site-packages/esmvalcore/_recipe/recipe.py", line 714, in __init__
    self.datasets = Dataset.from_recipe(recipe_file, session)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/projects/bsc32/software/suselinux/11/software/ESMValTool/2.10.0/lib/python3.11/site-packages/esmvalcore/dataset.py", line 152, in from_recipe
    return datasets_from_recipe(recipe, session)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/projects/bsc32/software/suselinux/11/software/ESMValTool/2.10.0/lib/python3.11/site-packages/esmvalcore/_recipe/to_datasets.py", line 386, in datasets_from_recipe
    variable_datasets = _get_datasets_for_variable(
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/projects/bsc32/software/suselinux/11/software/ESMValTool/2.10.0/lib/python3.11/site-packages/esmvalcore/_recipe/to_datasets.py", line 350, in _get_datasets_for_variable
    for facets, supplementaries in _get_facets_from_recipe(
  File "/gpfs/projects/bsc32/software/suselinux/11/software/ESMValTool/2.10.0/lib/python3.11/site-packages/esmvalcore/_recipe/to_datasets.py", line 327, in _get_facets_from_recipe
    yield _get_dataset_facets_from_recipe(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/projects/bsc32/software/suselinux/11/software/ESMValTool/2.10.0/lib/python3.11/site-packages/esmvalcore/_recipe/to_datasets.py", line 281, in _get_dataset_facets_from_recipe
    check.variable(
  File "/gpfs/projects/bsc32/software/suselinux/11/software/ESMValTool/2.10.0/lib/python3.11/site-packages/esmvalcore/_recipe/check.py", line 112, in variable
    f"{var['diagnostic']}")
       ~~~^^^^^^^^^^^^^^
KeyError: 'diagnostic'
2024-01-30 16:30:49,405 UTC [3133476] INFO    
If you have a question or need help, please start a new discussion on https://github.com/ESMValGroup/ESMValTool/discussions
If you suspect this is a bug, please open an issue on https://github.com/ESMValGroup/ESMValTool/issues
To make it easier to find out what the problem is, please consider attaching the files run/recipe_*.yml and run/main_log_debug.txt from the output directory.
```

### PR DESCRIPTION
**Pull request by enekomartinmartinez**
## Description

This PR should correct the bug of not correctly raising the error message in the variable check of the recipe.

Closes #2312

Link to documentation:

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by enekomartinmartinez** [comment]
I just updated the commit so as not to have the same error with the `variable_group`. It would be nice to include unity tests for these error messages.

File: esmvalcore/_recipe/check.py [hunk 1]
```diff
@@ -100,16 +100,21 @@ def duplicate_datasets(
...
     """Check variables as derived from recipe."""
     required = set(required_keys)
     missing = required - set(var)
     if missing:
         raise RecipeError(
             f"Missing keys {missing} in\n"
             f"{pformat(var)}\n"
-            "for variable {var['variable_group']} in diagnostic "
-            f"{var['diagnostic']}")
+            f"for variable {variable_group} in diagnostic "
```
Comment (bouweandela) [inline esmvalcore/_recipe/check.py:116]:
```suggestion
            f"for variable '{variable_group}' in diagnostic "
```

File: esmvalcore/_recipe/check.py [hunk 2]
```diff
@@ -100,16 +100,21 @@ def duplicate_datasets(
...
+            f"{diagnostic}")
```
Comment (bouweandela) [inline esmvalcore/_recipe/check.py:117]:
```suggestion
            f"'{diagnostic}'")
```

**Review by bouweandela — APPROVED** [review]
Thanks for taking the time to contribute! Looks good, just a minor suggestion to make the error message even more readable.

**Comment by bouweandela** [comment]
If you'd like to add a unit tests for this, see here for an example of how to write such a test: https://github.com/ESMValGroup/ESMValCore/blob/b0d30f65ccc34b9a85b3edcc722f77560fba2adf/tests/integration/recipe/test_check.py#L337

**Comment by enekomartinmartinez** [comment]
> Thanks for taking the time to contribute! Looks good, just a minor suggestion to make the error message even more readable.

Thanks for the suggestions, I will also update the error message in similar functions to have something more "uniform".

**Comment by bouweandela** [comment]
Would you still like to add a test, it shall we merge this as-is?

**Comment by enekomartinmartinez** [comment]
Hi @bouweandela I have tried to uniformize the RecipeErrors and tests them for:

- [diagnostics](https://github.com/ESMValGroup/ESMValCore/blob/6f30fa6495fbe76d5694e7ba9260e09db70ea631/esmvalcore/_recipe/check.py#L66)
- [duplicate_datasets](https://github.com/ESMValGroup/ESMValCore/blob/6f30fa6495fbe76d5694e7ba9260e09db70ea631/esmvalcore/_recipe/check.py#L90)
- [variable](https://github.com/ESMValGroup/ESMValCore/blob/6f30fa6495fbe76d5694e7ba9260e09db70ea631/esmvalcore/_recipe/check.py#L109)

Note that there are two errors (currently added a tests with `xfail`) that are not raised:
https://github.com/ESMValGroup/ESMValCore/blob/6f30fa6495fbe76d5694e7ba9260e09db70ea631/esmvalcore/_recipe/check.py#L71-L73
https://github.com/ESMValGroup/ESMValCore/blob/6f30fa6495fbe76d5694e7ba9260e09db70ea631/esmvalcore/_recipe/check.py#L84-L87

This is because the error was raised before by the [yamale.validate check](https://github.com/ESMValGroup/ESMValCore/blob/6f30fa6495fbe76d5694e7ba9260e09db70ea631/esmvalcore/_recipe/check.py#L57).

Therefore, for me, it would make sense to do any of these:
1. Remove the RecipeErrors that are not raised and convert the test to a yamale.validate errors
or
2. Soften the condition for the yamale.validate, setting `required=False` for those entries in the [recipe_schema.yml](https://github.com/ESMValGroup/ESMValCore/blob/6f30fa6495fbe76d5694e7ba9260e09db70ea631/esmvalcore/_recipe/recipe_schema.yml) and let the RecipeError raise later (which I think is more user friendly).

Let me know which one you prefer

**Comment by bouweandela** [comment]
I agree that option 2 is more user-friendly. Especially less technical users can get confused by the cryptic error messages coming from Yamale (see e.g. #1080 and #681)

**Comment by enekomartinmartinez** [comment]
@bouweandela, this is already okay and ready to merge by my side. Please, let me know if you find anything else missing!

**Comment by bouweandela** [comment]
Would you like to add yourself to the [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors)?

**Comment by enekomartinmartinez** [comment]
> Would you like to add yourself to the [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors)?

Done, thank you!

**Review by valeriupredoi — APPROVED** [review]
excellent work, gents :beers:

**Pull request merged**